### PR TITLE
[dbnode] Add reason tag to bootstrap retries metric

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Adds reason tag to bootstrap retries metric.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/storage/bootstrap_instrumentation.go
Comment (linasm) [inline src/dbnode/storage/bootstrap_instrumentation.go:171]: Have a look at `xerrors.Is(...)`, I think it takes care of nested errors.

Comment (linasm) [reply src/dbnode/storage/bootstrap_instrumentation.go:171]: I meant `Is` from `github.com/m3db/m3/src/x/errors`.

File: src/dbnode/storage/bootstrap_instrumentation_test.go
Comment (linasm) [inline src/dbnode/storage/bootstrap_instrumentation_test.go:86]: Nit:
```suggestion
			reason := ""
			if r, ok := counter.Tags()[reasonTag]; ok {
				reason = r
			}
			valuesByReason[reason] = int(counter.Value())
```

File: src/dbnode/storage/bootstrap_instrumentation_test.go
Comment (linasm) [inline src/dbnode/storage/bootstrap_instrumentation_test.go:35]: As I mentioned in one of the comments above, this test might be an overkill, I'd say testing this indirectly from `bootstrap_retries_test.go` is adequate. Leaving the final decision to you though.

Comment (vpranckaitis) [reply src/dbnode/storage/bootstrap_instrumentation_test.go]: Gonna leave, in case we add more retry `reason`s in the future.

File: src/dbnode/integration/bootstrap_retries_test.go
Comment (vpranckaitis) [reply src/dbnode/integration/bootstrap_retries_test.go:111]: I do not think it matters much if the assert is done during first bootstrapper pass, just before second pass or in between of them. As long as DB is not marked bootstrapped while bootstrapping is in progress (which should match the time period in between first and last `signalCh` writes, plus some variable amount of time before and after that), and as long as it's marked bootstrapped once bootstrapping is done, it should be fine. 

I guess I could make it so two signals are needed for bootstrapper to unblock. That way I could squeeze the assert in between them and that would guarantee exact execution on n-th bootstraper pass. Though I'm not sure if we would gain anything from this.

Comment (linasm) [reply src/dbnode/integration/bootstrap_retries_test.go:111]: But aren't you concerned that you could be advancing the clock at some unpredictable moment? Perhaps even before the first bootstrap pass has calculated the ranges to bootstrap?

Comment (linasm) [reply src/dbnode/integration/bootstrap_retries_test.go:113]: My bad, I was assuming the ranges are calculated within bs.Bootstrap(ctx, namespaces, cache) and not before invoking the anonymous fn.
Just make sure to rename noopNone to noopAll in places where it gets NewNoOpAllBootstrapperProvider assigned.

Comment (vpranckaitis) [reply src/dbnode/integration/bootstrap_retries_test.go:111]: Added double signal for more exact control when clock is advanced https://github.com/m3db/m3/pull/3317/commits/4181359ae40cec763adcde9e0adb99323b9be6cc
Refactored `noopNone` https://github.com/m3db/m3/pull/3317/commits/61967a8b256bd12d66fd70187ebec794ae77a25a

File: src/dbnode/storage/bootstrap_instrumentation_test.go
Comment (linasm) [reply src/dbnode/storage/bootstrap_instrumentation_test.go]: On average, writing code "for future" does not pay off :)

**Review by linasm — APPROVED** [review]
LGTM

**Pull request merged**
